rbac/default-role-manager/role_manager.go

```diff
@@ -329,9 +355,21 @@ func (rm *RoleManagerImpl) Clear() error {
 // AddLink adds the inheritance link between role: name1 and role: name2.
 // aka role: name1 inherits role: name2.
 func (rm *RoleManagerImpl) AddLink(name1 string, name2 string, domains ...string) error {
+	if hasLink, err := rm.HasLink(name2, name1, domains...); err != nil {
+		return err
+	} else if hasLink {
+		return fmt.Errorf("cannot add link from '%s' to '%s': would create a cycle", name1, name2)
+	}
```

Comment:
It is best not to use `HasLink` for loop detection, because it not only includes normal role inheritance, but also pattern matching, which does not match the logic of method `updateLevel`. I think it is better to use `visit[]` to mark in updateLevel